source/common/network/socket_impl.h

```diff
@@ -17,6 +17,11 @@ class SocketImpl : public virtual Socket {
     local_address_ = local_address;
   }
 
+  SocketPtr duplicate() override {
+    auto socket = new SocketImpl(io_handle_->duplicate(), local_address_);
+    return SocketPtr{socket};
```

Comment:
nit: return std::make_unique<SocketImpl>(...)

Reply:
I am not sure if this is actually possible because:
1. we both use a private constructor so the syntax `std::make_unique<SocketImpl>(io_handle_->duplicate(), local_address_)` does not compile
2. we have an deleted the copy constructor of `SocketImpl` so the syntax `std::make_unique<SocketImpl>(SocketImpl(io_handle_->duplicate(), local_address_));` does not compile

Reply:
I ran into ListenSocketImpl.  Moving the duplicate implementation to source/common/network/listen_socket_impl.h may solve the issue.

Reply:
I added on `SocketImpl` because I did not want to implement it for all types for sockets.

If I move the implementation to the `ListenSocketImpl` I would need to add a default implementation to all other sockets right?

Reply:
I think it would be acceptable for this call to fail for non listener sockets.

Reply:
What is the standard approach for these type of behavior? Do we have the equivalent of `NotImplementedException` or just add `RELEASE_ASSERT(false)`

Reply:
fixed in 933690154df8f498bcd10d66c195f991fcda46e2

Reply:
RELEASE_ASSERT(false, "not implemented") seems fine.

One thing we talked about offline was about ListenSocketImpl objects getting sliced when duplicated using this base class method that uses the SocketImpl constructor.  I think that's the main reason to move the duplicate implementation to ListenSocketImpl.